File: app/src/main/java/xlagunas/cat/andrtc/di/modules/ApplicationModule.java
package xlagunas.cat.andrtc.di.modules;

import android.app.Application;
import android.content.Context;

import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import xlagunas.cat.andrtc.UIThread;
import xlagunas.cat.andrtc.di.PerActivity;
import xlagunas.cat.data.cache.UserCache;
import xlagunas.cat.data.cache.UserCacheImpl;
import xlagunas.cat.data.repository.UserRepositoryImpl;
import xlagunas.cat.domain.executor.PostExecutionThread;
import xlagunas.cat.domain.interactor.LoginUseCase;
import xlagunas.cat.domain.interactor.UseCase;
import xlagunas.cat.domain.repository.UserRepository;

/**
 * Created by xlagunas on 29/02/16.
 */

@Module
public class ApplicationModule {

    private Application application;

    public ApplicationModule(Application app){
        this.application = app;
    }

    @Provides
    @Singleton
    public Context getAppContext(){
        return application;
    }

    @Provides @Singleton
    UserRepository provideUserRepository(UserRepositoryImpl userRepositoryImpl){
        return userRepositoryImpl;
    }

    @Provides @Singleton
    PostExecutionThread providePostExecutionThread(UIThread thread){
        return thread;
    }

    @Provides @Singleton
    UserCache provideUserCache(UserCacheImpl userCache) {
        return userCache;
    }

}

File: app/src/main/java/cat/xlagunas/viv/commons/di/ApplicationModule.kt
package cat.xlagunas.viv.commons.di

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import cat.xlagunas.data.common.converter.FriendConverter
import cat.xlagunas.data.common.converter.UserConverter
import cat.xlagunas.data.common.provider.ActivityMonitor
import cat.xlagunas.data.user.authentication.AuthTokenPreferenceDataStore
import cat.xlagunas.domain.schedulers.RxSchedulers
import cat.xlagunas.domain.user.authentication.AuthTokenDataStore
import dagger.Module
import dagger.Provides
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Singleton

@Module
class ApplicationModule {

    @Provides
    @Singleton
    fun provideAppContext(app: Application): Context = app

    @Provides
    @Singleton
    fun provideSharedPreference(context: Context): SharedPreferences =
            context.getSharedPreferences("preferences", Context.MODE_PRIVATE)

    @Provides
    @Singleton
    fun provideRxSchedulers() = RxSchedulers(Schedulers.io(), AndroidSchedulers.mainThread(), Schedulers.computation())

    @Provides
    @Singleton
    fun provideUserMapping() = UserConverter()

    @Provides
    @Singleton
    fun provideFriendMapping() = FriendConverter()

    @Provides
    @Singleton
    fun provideActivityMonitor() = ActivityMonitor()

    @Provides
    @Singleton
    fun provideAuthDataStore(sharedPreferences: SharedPreferences): AuthTokenDataStore = AuthTokenPreferenceDataStore(sharedPreferences)

}